### ISSUE TITLE
expected np.ndarray (got Tensor)

### ISSUE THREAD

**Issue by pbrandl** [pranaymanocha/PerceptualAudio#14]
I am passing tensors to the forward function and get following error:

expected np.ndarray (got Tensor)

When i transform to numpy I am loosing the gradient which I need for training.

Could you help me out here?

**Comment by pranaymanocha** [comment]
Hi,
I assume you are trying to use DPAM and face this error? I mention in the README that the current pip version supports evaluation, and not training using the metric. For training, you need to clone the repo and modify your code to be able to use it for training. ** However, you can use CDPAM for eval and training, using pip (based on pytorch) **
The current way this are setup for DPAM - its based on tf1.14 and the way I evaluate the graph, it returns a np array makes it impossible to be useful for training. I dont know of a nice and simple way to make it happen in tf 1.14, hence any suggestions would be great.

**Comment by pbrandl** [comment]
Hi,

I used the cdpam (installed via pip install cdpam). 

Then I exchanged the MSELoss for cdpam in pytorch.

My train function looks like this
`
loss_fn = cdpam.CDPAM()
prediction = model(x)
loss = loss_fn.forward(prediction, y)
`

`x` and `y` are tensors running on cuda. However, CDPAM forward expects numpy arrays which it then retransforms to tensors, if I am not mistaken. I will try to rewrite the code, such that it expects tensors :)

I think if you want to evaluate with numpy, you have to compute the backward way yourself. I don't think there is a simple trick...

**Comment by pbrandl** [comment]
Solved by simply rewriting the forward function to:

`
def forward(self, audio1=1, audio2=1):
        print("tet")

        _,a1,c1 = self.model.base_encoder.forward(audio1)
        a1 = F.normalize(a1, dim=1)
        _,a2,c2 = self.model.base_encoder.forward(audio2)
        a2 = F.normalize(a2, dim=1)
        dist1 = self.model.model_dist.forward(a1,a2)
`

**Comment by pranaymanocha** [comment]
Thanks for pointint it out! I have made the corrections and reuploaded the new pip version, and you can now pass tensors as inputs directly, as well as numpy arrays (for just eval).
